storage/blob.go

```diff
@@ -438,6 +498,154 @@ func (b BlobStorageClient) ContainerExists(name string) (bool, error) {
+// Returns permissionResponse which is combined permissions and AccessPolicy
+func (b BlobStorageClient) GetContainerPermissions(container string, timeout int, leaseID string) (permissionResponse *ContainerAccessResponse, err error) {
+	params := url.Values{"restype": {"container"},
+		"comp": {"acl"}}
+
+	if timeout > 0 {
+		params.Add("timeout", strconv.Itoa(timeout))
+	}
+
+	uri := b.client.getEndpoint(blobServiceName, pathForContainer(container), params)
+	headers := b.client.getStandardHeaders()
+
+	if leaseID != "" {
+		headers[leaseID] = leaseID
+	}
+
+	resp, err := b.client.exec("GET", uri, headers, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// containerAccess. Blob, Container, empty
+	containerAccess := resp.headers.Get(http.CanonicalHeaderKey(ContainerAccessHeader))
+
+	defer resp.body.Close()
+	var out AccessPolicy
+	err = xmlUnmarshal(resp.body, &out.SignedIdentifiersList)
+	if err != nil {
+		return nil, err
+	}
+
+	permissionResponse = &ContainerAccessResponse{}
+	permissionResponse.AccessPolicy = out.SignedIdentifiersList
+	permissionResponse.ContainerAccess = ContainerAccessType(containerAccess)
+
+	return permissionResponse, nil
+}
+
+func (b BlobStorageClient) generatePermissions(accessPolicy AccessPolicyDetails) (permissions string) {
+	// generate the permissions string (rwd).
+	// still want the end user API to have bool flags.
+	permissions = ""
+
+	if accessPolicy.CanRead {
+		permissions += "r"
+	}
+
+	if accessPolicy.CanWrite {
+		permissions += "w"
+	}
+
+	if accessPolicy.CanDelete {
+		permissions += "d"
+	}
+
+	return permissions
+}
+
+// convertAccessPolicyToXMLStructs converts between AccessPolicyDetails which is a struct better for API usage to the
+// AccessPolicy struct which will get converted to XML.
+func (b BlobStorageClient) convertAccessPolicyToXMLStructs(accessPolicy AccessPolicyDetails) (accessPolicyToMarshal AccessPolicy) {
+	accessPolicyToMarshal = AccessPolicy{}
+	accessPolicyToMarshal.SignedIdentifiersList = SignedIdentifiers{}
+	accessPolicyToMarshal.SignedIdentifiersList.SignedIdentifiers = []SignedIdentifier{}
+	si := SignedIdentifier{}
+	si.ID = accessPolicy.ID
+	si.AccessPolicy = AccessPolicyDetailsXML{}
+	si.AccessPolicy.StartTime = accessPolicy.StartTime.Round(time.Second)
+	si.AccessPolicy.ExpiryTime = accessPolicy.ExpiryTime.Round(time.Second)
+	permissions := b.generatePermissions(accessPolicy)
+	si.AccessPolicy.Permission = permissions
+	accessPolicyToMarshal.SignedIdentifiersList.SignedIdentifiers = append(accessPolicyToMarshal.SignedIdentifiersList.SignedIdentifiers, si)
+	return accessPolicyToMarshal
+}
+
+// generateAccessPolicy generates the XML access policy used as the payload for SetContainerPermissions.
+func (b BlobStorageClient) generateAccessPolicy(accessPolicy AccessPolicyDetails) (accessPolicyXML string, err error) {
+
+	if accessPolicy.ID != "" {
+		ap := b.convertAccessPolicyToXMLStructs(accessPolicy)
+		body, _, err := xmlMarshal(ap.SignedIdentifiersList)
```

Comment:
Changed. Was still trying to figure out XML Marshaling in Go. Still a strange strange beast to me :)